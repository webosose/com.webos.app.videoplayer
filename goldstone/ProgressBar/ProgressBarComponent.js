/**
 * Provides Progress Bar component as per the Y21 UX specs.
 *
 * @example
 * <ProgressBar progress={0.5} backgroundProgress={0.75} />
 *
 * @module goldstone/ProgressBar
 * @exports ProgressBar
 * @exports ProgressBarBase
 * @exports ProgressBarDecorator
 * @exports ProgressBarTooltip
 */
import ComponentOverride from '@enact/ui/ComponentOverride';
import UiProgressBar from './ProgressBarBase/ProgressBarBase';
import Slottable from '@enact/ui/Slottable';
import PropTypes from 'prop-types';
import compose from 'ramda/src/compose';
import React from 'react';
import generateClassNames from '../util/generateClassNames';
import Skinnable from '../Skinnable';
import {ProgressBarTooltip} from './ProgressBarTooltip';
import classNames from 'classnames';
import componentCss from './ProgressBarComponent.module.less';
/**
 * Renders a goldstone(based on Y21 and Sandstone)-styled progress bar.
 *
 * @function ProgressBar
 * @memberof goldstone/ProgressBar
 * @ui
 * @public
 */
const ProgressBarComponent = ({backgroundProgress, className, highlighted, orientation, disabled, progress, tooltip, ...rest}) => {
	const ProgressBarComponentClassNames = generateClassNames(componentCss, {
		highlighted,
		radial: (orientation === 'radial'),
		fillOverHalf: (progress > 0.5),
		loadOverHalf: (backgroundProgress > 0.5)
	});
	return (
		<UiProgressBar
			{...rest}
			disabled={disabled}
			backgroundProgress={backgroundProgress}
			orientation={orientation}
			progress={progress}
			css={componentCss}
			className={classNames({
				[className]: true,
				[componentCss.progressBar]: true,
				[ProgressBarComponentClassNames]: true
			})}
		>
			<ComponentOverride
				component={tooltip}
				orientation={orientation}
				percent
				proportion={progress}
				visible
			/>
		</UiProgressBar>
	);
};

ProgressBarComponent.propTypes = {
	backgroundProgress: PropTypes.number,
	disabled: PropTypes.bool,
	highlighted: PropTypes.bool,
	orientation: PropTypes.string,
	progress: PropTypes.number,
	progressAnchor: PropTypes.number,
	tooltip: PropTypes.bool
};

ProgressBarComponent.defaultProps = {
	/**
	 * The proportion of the loaded portion of the progress bar.
	 *
	 * * Valid values are between `0` and `1`.
	 *
	 * @type {Number}
	 * @default 0
	 * @public
	 */
	backgroundProgress: 0,
	/**
	 * Highlights the filled portion.
	 *
	 * @type {Boolean}
	 * @public
	 */
	highlighted: false,
	/**
	 * The orientation of the slider.
	 *
	 * @type {('horizontal'|'vertical'|'radial')}
	 * @default 'horizontal'
	 * @public
	 */
	orientation: 'horizontal',
	/**
	 * A number between `0` and `1` indicating the proportion of the filled portion of the bar.
	 *
	 * @type {Number}
	 * @default 0
	 * @public
	 */
	progress: 0,
	/**
	 * A number between `0` and `1` indicating the proportion of Progress bar to get Started(initial point).
	 *
	 * @type {Number}
	 * @default 0
	 * @public
	 */
	progressAnchor: 0,
	/**
	 * Enables the built-in tooltip.
	 *
	 * To customize the tooltip, pass either a custom tooltip component or an instance of
	 * [ProgressBarTooltip]{@link ProgressBar.ProgressBarTooltip} with additional
	 * props configured.
	 *
	 * The provided component will receive the following props from `ProgressBar`:
	 *
	 * * `orientation`  - The value of `orientation`
	 * * `percent`      - Always `true` indicating the value should be presented as a percentage
	 *                    rather than an absolute value
	 * * `progress`     - The `value` as a proportion between `min` and `max`
	 * * `visible`      - Always `true` indicating that the tooltip should be visible
	 *
	 * Usage:
	 * ```
	 * <ProgressBar
	 *   tooltip={
	 *     <ProgressBarTooltip position="after" />
	 *   }
	 * />
	 * ```
	 *
	 * The tooltip may also be passed as a child via the `"tooltip"` slot. See
	 * [Slottable]{@link ui/Slottable} for more information on how slots can be used.
	 *
	 * Usage:
	 * ```
	 * <ProgressBar>
	 *   <ProgressBarTooltip position="after" />
	 * </ProgressBar>
	 * ```
	 *
	 * @type {Boolean|Component|Element}
	 * @public
	 */
	tooltip: false
};
const ProgressBarDecorator = compose(
	Slottable({slots: ['tooltip']}),
	Skinnable
);
const ProgressBar = React.memo(ProgressBarDecorator(ProgressBarComponent));

export default ProgressBar;
export {
	ProgressBar,
	ProgressBarComponent,
	ProgressBarDecorator,
	ProgressBarTooltip
};
