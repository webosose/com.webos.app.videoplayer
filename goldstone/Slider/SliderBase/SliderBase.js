import EnactPropTypes from '@enact/core/internal/prop-types';
import Changeable from '@enact/ui/Changeable';
import ComponentOverride from '@enact/ui/ComponentOverride';
import Touchable from '@enact/ui/Touchable';
import PropTypes from 'prop-types';
import compose from 'ramda/src/compose';
import React from 'react';
import {KnobBase} from './Knob';
import PositionDecorator from './PositionDecorator';
import componentCss from './SliderBase.module.less';
import {calcProportion} from '../../util/calcProportion';
import classNames from 'classnames';
import generateClassNames from '../../util/generateClassNames';

const SliderBase = (props) => {
	const {
		backgroundProgress,
		css,
		disabled,
		orientation,
		noFill,
		pressed,
		progressBarComponent,
		progressAnchor,
		tooltipComponent,
		...rest
	} = props;
	delete rest.knobComponent;
	const calcPercentage = ({max, min, value = min}) => calcProportion(min, max, value);
	const calCulatedStyle = ({max, min, value = min}) => {
		let proportion = calcProportion(min, max, value);

		return {
			...rest.style,
			'--ui-slider-proportion-end-knob': proportion
		};
	};
	const appendedClassName = [componentCss.slider, generateClassNames(componentCss, {orientation, disabled, noFill, pressed}), generateClassNames(css, {orientation, disabled, noFill, pressed}), rest.className];
	delete rest.className;
	return (

		<div
			className={classNames(appendedClassName, componentCss.slider, css.slider)}
			style={calCulatedStyle(props)}
			disabled={disabled}
			{...rest}
		>

			<ComponentOverride
				className={classNames({[css.progressBar]: true, [componentCss.progressBar]: true})}
				backgroundProgress={backgroundProgress}
				component={progressBarComponent}
				orientation={orientation}
				progress={calcPercentage(props)}
				progressAnchor={progressAnchor}
			>
				<ComponentOverride
					className={classNames({[css.knob]: true, [componentCss.knob]: true})}
					component={KnobBase}
					disabled={disabled}
					orientation={orientation}
					proportion={calcPercentage(props)}
					tooltipComponent={tooltipComponent}
					value={props.value}
				/>
			</ComponentOverride>
		</div >
	);

};

SliderBase.propTypes = {
	progressBarComponent: EnactPropTypes.componentOverride.isRequired,
	backgroundProgress: PropTypes.number,
	css: PropTypes.object,
	disabled: PropTypes.bool,
	knobComponent: EnactPropTypes.componentOverride,
	max: PropTypes.number,
	min: PropTypes.number,
	noFill: PropTypes.bool,
	orientation: PropTypes.oneOf(['horizontal', 'vertical']),
	pressed: PropTypes.bool,
	progressAnchor: PropTypes.number,
	step: PropTypes.number,
	tooltipComponent: EnactPropTypes.componentOverride,
	value: PropTypes.number

};
SliderBase.defaultProps = {
	disabled: false,
	knobComponent: KnobBase,
	min: 0,
	max: 100,
	noFill: false,
	orientation: 'horizontal',
	progressAnchor: 0,
	step: 1
};
const SliderDecorator = compose(
	Changeable,
	PositionDecorator,
	Touchable({activeProp: 'pressed'})
);
const Slider = SliderDecorator(SliderBase);

export default Slider;

export {
	KnobBase,
	Slider,
	SliderBase,
	SliderDecorator
};
