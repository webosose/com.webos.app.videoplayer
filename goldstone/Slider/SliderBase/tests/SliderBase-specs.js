import ComponentOverride from '@enact/ui/ComponentOverride';
import React from 'react';
import {mount} from 'enzyme';
import {SliderBase} from './../SliderBase';
import ProgressBar from './../../../ProgressBar/ProgressBarBase/ProgressBarBase';
import componentCss from '../ProgressBarComponent.module.less';
import {ProgressBarTooltip} from './../../../ProgressBar';

describe('SliderBase component', () => {

	it('SliderBase renders as expected!!!', () => {
		const wrapper = mount(
			<SliderBase
				disabled={false}
				css={componentCss}
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);
		expect(wrapper).toMatchSnapshot();
	});

	it('the sliderbar is disabled ', () => {
		const wrapper = mount(
			<SliderBase
				disabled={false}
				css={componentCss}
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);

		const value = wrapper.find('.slider').prop('disabled');
		expect(value).toBeFalsy();
	});

	it('the sliderbar is enable ', () => {
		const wrapper = mount(
			<SliderBase
				disabled
				css={componentCss}
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);

		const value = wrapper.find('.slider').prop('disabled');
		expect(value).toBeTruthy();
	});

	it('the orientation is horizontal ', () => {
		const wrapper = mount(
			<SliderBase
				disabled
				css={componentCss}
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);

		const value = wrapper.find('SliderBase').prop('orientation');
		expect(value).toBe('horizontal');
	});
	it('the orientation is vertical', () => {
		const wrapper = mount(
			<SliderBase
				disabled
				css={componentCss}
				orientation={'vertical'}
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);

		const value = wrapper.find('SliderBase').prop('orientation');
		expect(value).toBe('vertical');
	});

	it('the noFill value is false', () => {
		const wrapper = mount(
			<SliderBase
				disabled
				css={componentCss}
				orientation={'vertical'}
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);

		const value = wrapper.find('SliderBase').prop('noFill');
		expect(value).toBeFalsy();
	});


	it('the noFill value is true', () => {
		const wrapper = mount(
			<SliderBase
				disabled
				css={componentCss}
				orientation={'vertical'}
				noFill
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);

		const value = wrapper.find('SliderBase').prop('noFill');
		expect(value).toBeTruthy();
	});

	it('the min value is zero by default', () => {
		const wrapper = mount(
			<SliderBase
				disabled
				css={componentCss}
				orientation={'vertical'}
				noFill
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);

		const value = wrapper.find('SliderBase').prop('min');
		expect(value).toBe(0);
	});

	it('the max value is 100 by default', () => {
		const wrapper = mount(
			<SliderBase
				disabled
				css={componentCss}
				orientation={'vertical'}
				noFill
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);

		const value = wrapper.find('SliderBase').prop('max');
		expect(value).toBe(100);
	});

	it('the max value is 50', () => {
		const wrapper = mount(
			<SliderBase
				disabled
				css={componentCss}
				orientation={'vertical'}
				max={50}
				noFill
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);

		const value = wrapper.find('SliderBase').prop('max');
		expect(value).toBe(50);
	});

	it('the min value is 50', () => {
		const wrapper = mount(
			<SliderBase
				disabled
				css={componentCss}
				orientation={'vertical'}
				min={50}
				noFill
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);

		const value = wrapper.find('SliderBase').prop('min');
		expect(value).toBe(50);
	});

	it('default value of progressAnchor is zero ', () => {
		const wrapper = mount(
			<SliderBase
				disabled
				css={componentCss}
				orientation={'vertical'}
				min={50}
				noFill
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);

		const value = wrapper.find('SliderBase').prop('progressAnchor');
		expect(value).toBe(0);
	});

	it('the value of progressAnchor is 50 ', () => {
		const wrapper = mount(
			<SliderBase
				disabled
				css={componentCss}
				orientation={'vertical'}
				progressAnchor={50}
				min={50}
				noFill
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);

		const value = wrapper.find('SliderBase').prop('progressAnchor');
		expect(value).toBe(50);
	});

	it('by default the value of step is 1 ', () => {
		const wrapper = mount(
			<SliderBase
				disabled
				css={componentCss}
				orientation={'vertical'}
				progressAnchor={50}
				min={50}
				noFill
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);

		const value = wrapper.find('SliderBase').prop('step');
		expect(value).toBe(1);
	});

	it('by default the value of step is 50 ', () => {
		const wrapper = mount(
			<SliderBase
				disabled
				css={componentCss}
				orientation={'vertical'}
				progressAnchor={50}
				min={50}
				noFill
				step={50}
				progressBarComponent={
					<ProgressBar
						css={componentCss}
					/>
				}
				tooltipComponent={
					<ComponentOverride
						component={ProgressBarTooltip}
						visible
					/>
				}
			/>


		);

		const value = wrapper.find('SliderBase').prop('step');
		expect(value).toBe(50);
	});

});
