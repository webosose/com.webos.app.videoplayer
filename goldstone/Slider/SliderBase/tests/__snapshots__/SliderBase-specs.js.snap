// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SliderBase component SliderBase renders as expected!!! 1`] = `ReactWrapper {}`;
